### ISSUE TITLE
Some operators incorrectly cache count of a collection source

### ISSUE THREAD

**Issue by viceroypenguin** [morelinq/MoreLINQ#943]
In some places (see `CountDown+IterateCollection`), `TryGetCollectionCount` is used incorrectly. The `Count` of the collection is evaluated at the time the enumerable is created, rather than when it is enumerated. However, the length of the collection could change in between the time the enumeration is created and the time it is enumerated. As such, the value of collection `Count` should be obtained at the time of enumeration. `TryGetCollectionCount` can be used to determine that a `source` is a collection for the purposes of determining which method to call; but if done, the value should not be relied on.

List of potentially affected operators (added by @atifaziz):

- [x] `AssertCount`
- [x] `Batch`
- [x] `CountDown`
- [x] `FallbackIfEmpty`
- [x] `PadStart`
- [x] `SkipLast`
- [x] `TakeLast`

**Comment by atifaziz** [comment]
This is a good find and should be addressed! Thanks for reporting it!

Demo for `CountDown`:

```c#
var stack = new Stack<int>(Enumerable.Range(1, 3));
var q = stack.CountDown(1, (_, cd) => cd);
stack.Push(4);
foreach (var cd in q)
    Console.WriteLine(cd);
```

Outputs:

    null
    null
    0
    -1

**Comment by atifaziz** [comment]
Closing now that PR #965 is merged with 0addfbd5673d0d04056709b62a6ad03ca4ba5315.

Thanks again @viceroypenguin!
